Resources/public/js/views/services/plugins/ez-locationswapplugin.js

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) eZ Systems AS. All rights reserved.
+ * For full copyright and license information view LICENSE file distributed with this source code.
+ */
+YUI.add('ez-locationswapplugin', function (Y) {
+    "use strict";
+    /**
+     * Provides the plugin for swap location
+     *
+     * @module ez-locationswapplugin
+     */
+    Y.namespace('eZ.Plugin');
+
+    /**
+     * Swap location plugin. It sets an event handler to swap location
+     *
+     * In order to use it you need to fire `swapLocation` event with parameter
+     * `location` containing the eZ.Location object you want swap with other.
+     *
+     * @namespace eZ.Plugin
+     * @class LocationSwap
+     * @constructor
+     * @extends eZ.Plugin.ViewServiceBase
+     */
+    Y.eZ.Plugin.LocationSwap = Y.Base.create('locationswapplugin', Y.eZ.Plugin.ViewServiceBase, [], {
+
+        initializer: function () {
+            this.onHostEvent('*:swapLocation', this._selectDestinationLocation);
+        },
+
+        /**
+         * swapLocation event handler, launch the universal discovery widget
+         * to choose the location you want to swap with the given one
+         *
+         * @method _selectDestinationLocation
+         * @private
+         * @param {EventFacade} e swapLocation event facade
+         */
+        _selectDestinationLocation: function (e) {
+            var service = this.get('host'),
+                originLocationId = e.location.get('id'),
+                originLocationCanHaveChildren = service.get('contentType').get('isContainer'),
+                originLocationHasChildren = originLocationCanHaveChildren &&
+                    (service.get('location').get('childCount') > 0);
+
+            service.fire('contentDiscover', {
+                config: {
+                    title: "Select the location you want to swap with the given one",
+                    contentDiscoveredHandler: Y.bind(this._swapLocations, this),
```

Comment:
you have to pass the *originLocation* (ie `e.location`) in the parameter of this function and use it in `_swapLocations` instead of `service.get('location')` to make sure the Location provided in the event is always used as the *source*, so here you can write:
```js
Y.bind(this._swapLocations, this, e.location),
```

and the `_swapLocations` signature becomes:
```js
_swapLocations: function (location, e) {
```